flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnApplicationClusterEntryPoint.java

```diff
@@ -0,0 +1,196 @@
+
+	public static void main(final String[] args) {
+		// startup checks and logging
+		EnvironmentInformation.logEnvironmentInfo(LOG, YarnApplicationClusterEntryPoint.class.getSimpleName(), args);
+		SignalHandler.register(LOG);
+		JvmShutdownSafeguard.installAsShutdownHook(LOG);
+
+		Map<String, String> env = System.getenv();
+
+		final String workingDirectory = env.get(ApplicationConstants.Environment.PWD.key());
+		Preconditions.checkArgument(
+				workingDirectory != null,
+				"Working directory variable (%s) not set",
+				ApplicationConstants.Environment.PWD.key());
+
+		try {
+			YarnEntrypointUtils.logYarnEnvironmentInformation(env, LOG);
+		} catch (IOException e) {
+			LOG.warn("Could not log YARN environment information.", e);
+		}
+
+		final Configuration configuration = YarnEntrypointUtils.loadConfiguration(workingDirectory, env);
+		final ApplicationConfiguration applicationConfiguration = ApplicationConfiguration.fromConfiguration(configuration);
+
+		PackagedProgram program = null;
+		try {
+			program = getPackagedProgram(configuration, applicationConfiguration);
+		} catch (Exception e) {
+			LOG.error("Could not create application program.", e);
+			System.exit(1);
+		}
+
+		final JobID  jobId = createJobIdForCluster(configuration);
+		configuration.set(PipelineOptionsInternal.PIPELINE_FIXED_JOB_ID, jobId.toHexString());
+
+		configuration.set(DeploymentOptions.TARGET, EmbeddedExecutor.NAME);
+		ConfigUtils.encodeCollectionToConfig(configuration, PipelineOptions.JARS, program.getJobJarAndDependencies(), URL::toString);
+		ConfigUtils.encodeCollectionToConfig(configuration, PipelineOptions.CLASSPATHS, program.getClasspaths(), URL::toString);
+
+		YarnApplicationClusterEntryPoint yarnApplicationClusterEntrypoint =
+				new YarnApplicationClusterEntryPoint(configuration, program);
+
+		ClusterEntrypoint.runClusterEntrypoint(yarnApplicationClusterEntrypoint);
+	}
+
+	private static PackagedProgram getPackagedProgram(
+			final Configuration configuration,
+			final ApplicationConfiguration applicationConfiguration) throws IOException, FlinkException {
+		final PackagedProgramRetriever programRetriever = getPackagedProgramRetriever(
+				configuration,
+				applicationConfiguration.getProgramArguments(),
+				applicationConfiguration.getApplicationClassName());
+		return programRetriever.getPackagedProgram();
+	}
+
+	private static PackagedProgramRetriever getPackagedProgramRetriever(
+			final Configuration configuration,
+			final String[] programArguments,
+			@Nullable final String jobClassName) throws IOException {
+		final ClassPathPackagedProgramRetriever.Builder retrieverBuilder =
+				ClassPathPackagedProgramRetriever
+						.newBuilder(programArguments)
+						.setJobClassName(jobClassName);
+		getUsrLibDir(configuration).ifPresent(retrieverBuilder::setUserLibDirectory);
+		return retrieverBuilder.build();
+	}
+
+	private static Optional<File> getUsrLibDir(final Configuration configuration) {
```

Comment:
We could put `getUsrLibDir` into the `YarnEntrypointUtils` to avoid duplication with `YarnJobClusterEntrypoint`.